### PR TITLE
ci: Compile Native Modules via Actions

### PR DESCRIPTION
**Pull request by BradenM**
This PR adds a Github actions workflow to automate compiling native modules for pymakr-vsc.

The check suites can be viewed [here](https://github.com/BradenM/pymakr-vsc/actions).

## Overview:

- [x] Retrieve tags from vscode git repository
- [x] Fetch and parse the Electron target runtime for each tag
- [x] Execute `electron-rebuild` to rebuild native modules for the latest and past 3 target runtimes
- [ ] Test each binding to ensure they work properly
- [x] Use `mp-download.ps1` to harvest the bindings (adding them to `native_modules`)
- [ ] Merge results from each environment and commit to `master` branch

## Notes

Currently all seems well for linux/darwin environments, but I am having a strange issue (seen [here](https://github.com/BradenM/pymakr-vsc/commit/4411f26d55388a1355ea81d67d1886ebfb664896/checks?check_suite_id=306288586) on the `windows-latest` environment in which `npm install` exits with an error code of `1` for no apparent reason. @Josverl, do you happen to have any ideas towards this? Windows is not my strong suit.

This is heading towards what I believe we are trying to accomplish, but please correct me if I am steering off course. I am not very comfortable with Powershell, so if I unnecessarily re-implemented some items feel free to point it out. 

Also, I will likely rebase some of these commits, since most of them are tiny. I can't seem to find a solid method of testing Git actions besides pushing to remote...

### CODE REVIEW

**Comment by Josverl** [comment]
so , 
- windows does not come loaded with compilers ( these are used before shipping the code :-) 
- node-gyph requires a compiler ( on any plaform)
- for windows : 
- to install them there is this NPM package that installs the C++ compiler 
https://github.com/nodejs/node-gyp#option-2
- and that also install python2.x (old stuff , was needed before the summer ) and python 2.7 installation for 2.7 is not too reliable to begin with  

long story short , it takes 40-80-?? minutes to install these things , and if they fail: cry , wipe , retry .
for now still waiting :
image

**Comment by BradenM** [comment]
@Josverl,

Yeah I actually tried adding `windows-build-tools` in an earlier commit, but ran into the same issue. I did find [this issue](https://github.com/felixrieseberg/windows-build-tools/issues/152#issuecomment-442488180) that suggests running some `choco` upgrade(s) before installing it. Currently trying it [here](https://github.com/BradenM/pymakr-vsc/commit/c4d2d34715f5752eafcc328ac81c220e237c9049/checks?check_suite_id=307795602).

**Comment by Josverl** [comment]
🤦🏻‍♂️ sorry, I did not see that you tried this already .
( I needed to squash in order to copy over you branch ...) 

Its probable the python 2.x install that fails

but isf pycomm moves to the new serialport , then then we can just download the binadings , rather then building them.
that is why i built mp-download in the first place .
( e.g. why compile if someone else already did that ....)

**Comment by BradenM** [comment]
Oh wow. Somehow I completely skimmed over that fact, haha. So do we even really need a compile action? Or does upgrading serialport lead to new issues?

Is this really as simple as -> upgrade serial port -> execute `mp-download.ps1` -> package ?

**Comment by Josverl** [comment]
well , I may have hardcoded/defaulted some serialnumbers 😎,But yeah

**Comment by Josverl** [comment]
and the testing needs to be done on each platform, so we still need the CI/CD for testing

**Comment by Josverl** [comment]
and if we would like to support muscl versions on linux , or arm versions of windows then i think there are no prebuilts published for these. 
but I think that is a stretch goal

**Comment by BradenM** [comment]
Okay, wow, big mistake on my part haha. Not sure how I utterly missed that, but alright.

Does this include everything for now?

1.) Setup package (checkout/npm install)
2.) Execute `mp-download` to update bindings
3.) Test bindings (via your `electron-serialport` or `checkABI.js` I assume, havn't looked to much into this yet)
4.) Run  tests: `npm run test` 
5.) Cleanup: `npm ci && npm prune --production`
6.) Package: `vsce package`
7.) **If pushed on a tag such as release/X.X.X** -> `vsce publish`

Done?

**Comment by Josverl** [comment]
yes I think so :
a few tweaks may be needed : 
a)  an minor but needed update to mp-download :
https://github.com/Josverl/pymakr-vsc/commit/ce0d65808b824367e3d283c3c61c06816b18f95a

b) I need to check the install scrips, (I think Pedro undid/replaced the install script in order to compile )

c) I would suggest to move ( or repeat) 4) tests until after 5) Cleanup 
the vsix package would be a 'nightly' / weekly build , so probably need to rename it as such 

d) and , is there a way to run this on a daily/weekly interval ?
that way we can test for updates to vscode even if no changes are made to pymakr

if not we can always run MP-Download with -electronversions '7.1.1' 
to hardcode a future version
